src/peep/peep.c

```diff
@@ -3421,7 +3478,785 @@ static void peep_update_fixing(int steps, rct_peep* peep){
+			break;
+
+		case 11:
+			loc_6C13F8(steps, peep, ride);
+			break;
+
+		case 12:
+			loc_6C14E5(steps, peep, ride);
+			break;
+
+		case 13:
+			loc_6C1559(steps, peep, ride);
+			break;
+
+		case 14:
+			loc_6C15EF(steps, peep, ride);
+			break;
+
+		default:
+			log_error("Invalid substate");
+			RCT2_CALLPROC_X(RCT2_ADDRESS(0x006C0EB0, uint32)[peep->sub_state], steps, 0, 0, 0, (int) peep, (int) ride, 0);
+	}
+
+}
+
+void loc_6C0EEC(int steps, rct_peep *peep, rct_ride *ride) {
+	// eax: steps
+	// esi: peep
+	// edi: ride
+
+	ride->mechanic_status = 3;
+	ride->window_invalidate_flags |= RIDE_INVALIDATE_RIDE_MAINTENANCE;
+
+	loc_6C1653(steps, peep, ride);
+}
+
+void loc_6C0F09(int steps, rct_peep *peep, rct_ride *ride) {
+	uint32 ebx = ride->broken_vehicle;
+	uint16 bp = ride->vehicles[ebx];
+
+	if (bp == 0xFFFF) {
+		loc_6C1653(steps, peep, ride);
+		return;
+	}
+
+	rct_vehicle *vehicle = GET_VEHICLE(bp);
+	sint8 ah = ride->broken_car;
+
+	// loc_6C0F3E:
+	while (true) {
+		ah--;
+		if (ah < 0) {
+			break;
+		}
+
+		vehicle = GET_VEHICLE(vehicle->next_vehicle_on_train);
+	}
+
+	// loc_6C0F51:
+	while (true) {
+		if (vehicle->is_child == 0) {
+			break;
+		}
+
+		uint8 trackType = vehicle->track_type >> 2;
+		if (trackType == TRACK_ELEM_END_STATION) {
+			break;
+		}
+
+		if (trackType == TRACK_ELEM_BEGIN_STATION) {
+			break;
+		}
+
+		if (trackType == TRACK_ELEM_MIDDLE_STATION) {
+			break;
+		}
+
+		vehicle = GET_VEHICLE(vehicle->prev_vehicle_on_ride);
+	}
+
+	// loc_6C0F7A:
+	uint32 peepDirection = peep->var_78;
+
+	sint16 ax = word_981D6C[peepDirection].x;
+	sint16 cx = word_981D6C[peepDirection].y;
+
+	ax *= -12;
+	cx *= -12;
+
+	ax += vehicle->x;
+	cx += vehicle->y;
+
+	peep->destination_x = ax;
+	peep->destination_y = cx;
+	peep->destination_tolerence = 2;
+
+	loc_6C0FAA(steps, peep, ride);
+}
+
+static void loc_6C0FAA(int steps, rct_peep *peep, rct_ride *ride) {
+	// eax: steps
+	// esi: peep
+	// edi: ride
+
+	invalidate_sprite_2((rct_sprite *) peep);
+	int out = peep_update_action(); // TODO: arguments
+	if (out == 0) {
+		uint16 dx = peep->z;
+		sprite_move(); // TODO: arguments
+		invalidate_sprite_2((rct_sprite *) peep);
+		return;
+	}
+
+	loc_6C1653(steps, peep, ride);
+}
+
+void loc_6C0FD3(int steps, rct_peep *peep, rct_ride *ride) {
+	peep->sprite_direction = peep->var_78 << 3;
+	scenario_rand();
+	peep->action = PEEP_ACTION_STAFF_FIX;
+	if (steps & 0x100000) {
+		peep->action = PEEP_ACTION_STAFF_FIX_2;
+	}
+
+	// loc_6C0FFB:
+	peep->action_sprite_image_offset = 0;
+	peep->action_frame = 0;
+	sub_693B58(peep);
+	invalidate_sprite_2((rct_sprite *) peep);
+
+	loc_6C100D(steps, peep, ride);
+}
+
+static void loc_6C100D(int steps, rct_peep *peep, rct_ride *ride) {
+	// eax: steps
+	// esi: peep
+	// edi: ride
+
+	if (peep->action == 0xFF) {
+		loc_6C1653(steps, peep, ride);
+		return;
+	}
+
+	peep_update_action(); // TODO: arguments
+
+	uint8 al = peep->action;
+	sint8 ah = 0x25;
+	if (al != 0x0F) {
```

Comment:
If PEEP_ACTION_STAFF_FIX then the frame that we fix the ride is different to STAFF_FIX_2